src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,181 @@
+            Node<T> newNode = new Node<>(currentNode, e, null);
+            if (currentNode == null) {
+                head = newNode;
+            } else {
+                currentNode.next = newNode;
+            }
+            currentNode = newNode;
+            size++;
+        }
+        tail = currentNode;
     }
 
     @Override
     public T get(int index) {
+        indexOnRangeEqualsSize(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        if (currentNode != null) {
+            return currentNode.value;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexOnRangeEqualsSize(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexOnRangeEqualsSize(index);
+        T removedData;
+        if (index == 0) {
+            removedData = head.value;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prevNode = getNodeByIndex(index - 1);
+            removedData = prevNode.next.value;
+            prevNode.next = prevNode.next.next;
+            if (index == size - 1) {
+                tail = prevNode;
+            }
+        }
+        size--;
+        return removedData;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.value == object || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> currentNode) {
+        Node<T> prevNode = currentNode.prev;
+        Node<T> nextNode = currentNode.next;
+        if (prevNode == null) {
+            head = nextNode;
+        } else {
+            prevNode.next = nextNode;
+            currentNode.prev = null;
+        }
+        if (nextNode == null) {
+            tail = prevNode;
+        } else {
+            nextNode.prev = prevNode;
+            currentNode.next = null;
+        }
+        size--;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void indexOnRange(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+    }
+
+    private void indexOnRangeEqualsSize(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
     }
```

Comment:
We can have 1 method, they are the same. Just pass `size` or `size + 1` as a parameter to your `indexOnRange`

Reply:
i just add one if into the add() method, thats ok?

Reply:
I think that is acceptable since it is actually a only single use case